### PR TITLE
fix: revive torch backend

### PR DESCRIPTION
**Pull request by roelofvandijk**
as few changes as possible - all tests in the CI pass again

added more test coverage to `test.py` (nice for development, can remove)

Caveats:
* had to jump through some hoops for circular padding and `Tensor.diagonal`, those are at the end of the file. Circular padding has a custom implementaion and explicit autograd registration - can't figure out a simpler way for now. 
* related old circpad PR:  https://github.com/tinygrad/tinygrad/pull/9642
* related old diag PR: https://github.com/tinygrad/tinygrad/pull/10945

I suggest fixing both in a different PR. I tried a lot of different variants for diag, but it's tricky. I suspect maybe `as_strided` not supporting diagonal strides or so.

For _as_strided, I just updated the simple version that copies (which works). Then I tried to split out the pure view options in https://github.com/tinygrad/tinygrad/pull/13280/commits/82a61223f2067ca8fa9de745c1938a4244e5a710. Happy to revert that.

### CODE REVIEW

**Comment by geohot** [comment]
There's way too many kernels for resnet. This has to be reasonable, it should retain the laziness and not be realizing at each layer.

**Comment by geohot** [comment]
Screenshot 2025-11-14 at 1 37 26 PM

**Comment by roelofvandijk** [comment]
1059 lines [kernels_cc2dfe2.txt](https://github.com/user-attachments/files/23560182/kernels_cc2dfe2.txt)
793 lines [kernels_current.txt](https://github.com/user-attachments/files/23560183/kernels_current.txt)
Before you removed the shapetracker (cc2dfe2, which is 2 commits before your "remove ShapeTracker"), this generated ~639 kernels, this PR is ~417. Should be far less, but that issue was already present with the ShapeTracker view tracking.

`PYTHONPATH=. DEBUG=2 python3 extra/torch_backend/example.py > kernels_current.txt`
`git checkout cc2dfe2  && PYTHONPATH=. DEBUG=2 python3 extra/torch_backend/example.py > kernels_cc2dfe2.txt`

Did this break somewhere before removal?

I am happy to dive into it more, there are probably realizes/contiguous that can go for better fusion, but that will mean a larger diff to the original backend.


```
(.venv) ➜  tinygrad git:(cc2dfe22f) ✗ grep -c "^\*\*\* METAL " kernels_cc2dfe2.txt
639
```
```
(.venv) ➜  tinygrad git:(cc2dfe22f) ✗ grep -c "^\*\*\* METAL " kernels_current.txt                                                              
417
```

**Comment by roelofvandijk** [comment]
Made some progress. One thing that can make a big difference is a custom batch norm wrapper to use `Tensor.batchnorm` instead of `  aten.native_batch_norm, aten.native_batch_norm_backward`, which can bring the kernels down to 235.

**Comment by geohot** [comment]
Custom kernels are not the solution and not something we want. The fusion is broken. Fix the fusion.

The issue is copies, probably `For _as_strided, I just updated the simple version that copies`

**Comment by roelofvandijk** [comment]
Resnet fuses, all tests pass, including some extra tests and some kernel fusion regression tests. I used these during development and added them to the CI, can remove them again if you want. Did not seem unreasonably slow (15s).


Resnet example now asserts on the expected number of kernels. I could add an expectation to the mnist example as well for batchnorm fusion during training.


```*** CPU      207 r_4_112_28_4_4_4_3_7_7                         arg  7 mem   0.05 GB tm     13.93ms/    40.26ms (     17 GFLOPS    0|13     GB/s) ['add', 'rsqrt', 'conv2d', 'batchnorm', 'relu']
*** CPU      208 E_250_4                                        arg  2 mem   0.05 GB tm     13.12us/    40.27ms (      0 GFLOPS    1|1      GB/s) 
*** CPU      209 r_16_56_56_4_3_3                               arg  2 mem   0.05 GB tm    425.36us/    40.70ms (     11 GFLOPS    9|19     GB/s) ['max_pool2d']
*** CPU      210 r_4_56_14_4_4_4_64_3_3                         arg  7 mem   0.05 GB tm   7067.15us/    47.76ms (     33 GFLOPS    0|25     GB/s) ['add', 'rsqrt', 'conv2d', 'batchnorm', 'relu']
*** CPU      211 r_4_56_14_4_4_4_64_3_3n1                       arg  8 mem   0.05 GB tm   7160.42us/    54.92ms (     32 GFLOPS    0|24     GB/s) ['add', 'rsqrt', 'conv2d', 'batchnorm', '__add__', 'relu']
*** CPU      212 r_4_56_14_4_4_4_64_3_3n2                       arg  7 mem   0.05 GB tm   7127.24us/    62.05ms (     33 GFLOPS    0|24     GB/s) ['add', 'rsqrt', 'conv2d', 'batchnorm', 'relu']
*** CPU      213 r_4_56_14_4_4_4_64_3_3n1                       arg  8 mem   0.05 GB tm   6847.13us/    68.90ms (     34 GFLOPS    0|26     GB/s) ['add', 'rsqrt', 'conv2d', 'batchnorm', '__add__', 'relu']
*** CPU      214 r_8_28_7_4_4_4_64_3_3                          arg  7 mem   0.05 GB tm   2746.16us/    71.64ms (     42 GFLOPS    1|37     GB/s) ['add', 'rsqrt', 'conv2d', 'batchnorm', 'relu']
*** CPU      215 r_8_28_7_4_4_4_128_3_3_16_4                    arg 13 mem   0.05 GB tm     17.34ms/    88.98ms (     14 GFLOPS    0|14     GB/s) ['add', 'rsqrt', 'conv2d', 'batchnorm', '__add__', 'relu']
*** CPU      216 r_8_28_7_4_4_4_128_3_3                         arg  7 mem   0.05 GB tm   7042.65us/    96.03ms (     33 GFLOPS    0|25     GB/s) ['add', 'rsqrt', 'conv2d', 'batchnorm', 'relu']
*** CPU      217 r_8_28_7_4_4_4_128_3_3n1                       arg  8 mem   0.05 GB tm   6930.59us/   102.96ms (     33 GFLOPS    0|25     GB/s) ['add', 'rsqrt', 'conv2d', 'batchnorm', '__add__', 'relu']
*** CPU      218 r_16_14_14_4_4_128_3_3                         arg  7 mem   0.05 GB tm   8619.37us/   111.57ms (     13 GFLOPS    0|34     GB/s) ['add', 'rsqrt', 'conv2d', 'batchnorm', 'relu']
*** CPU      219 r_16_14_14_4_4_256_3_3_32_4                    arg 13 mem   0.05 GB tm     20.81ms/   132.38ms (     12 GFLOPS    0|29     GB/s) ['add', 'rsqrt', 'conv2d', 'batchnorm', '__add__', 'relu']
*** CPU      220 r_16_14_14_4_4_256_3_3                         arg  7 mem   0.05 GB tm     18.60ms/   150.98ms (     12 GFLOPS    0|31     GB/s) ['add', 'rsqrt', 'conv2d', 'batchnorm', 'relu']
*** CPU      221 r_16_14_14_4_4_256_3_3n1                       arg  8 mem   0.05 GB tm     19.53ms/   170.52ms (     12 GFLOPS    0|30     GB/s) ['add', 'rsqrt', 'conv2d', 'batchnorm', '__add__', 'relu']
*** CPU      222 r_128_4_7_7_256_3_3                            arg  7 mem   0.05 GB tm   4448.07us/   174.97ms (     24 GFLOPS    1|24     GB/s) ['add', 'rsqrt', 'conv2d', 'batchnorm', 'relu']
*** CPU      223 r_128_4_7_7_512_3_3_64_4                       arg 13 mem   0.05 GB tm     12.33ms/   187.30ms (     20 GFLOPS    1|40     GB/s) ['add', 'rsqrt', 'conv2d', 'batchnorm', '__add__', 'relu']
*** CPU      224 r_128_4_7_7_512_3_3                            arg  7 mem   0.05 GB tm   8323.17us/   195.62ms (     23 GFLOPS    1|7      GB/s) ['add', 'rsqrt', 'conv2d', 'batchnorm', 'relu']
*** CPU      225 r_128_4_7_7_512_3_3n1                          arg  8 mem   0.05 GB tm   8260.66us/   203.88ms (     23 GFLOPS    1|7      GB/s) ['add', 'rsqrt', 'conv2d', 'batchnorm', '__add__', 'relu']
*** CPU      226 r_128_4_49                                     arg  2 mem   0.05 GB tm     21.44us/   203.91ms (      1 GFLOPS    5|5      GB/s) ['mean']
*** CPU      227 r_500_2_128_4                                  arg  3 mem   0.05 GB tm    358.61us/   204.26ms (      3 GFLOPS    6|11     GB/s) ['matmul']
*** CPU      228 E_250_4                                        arg  2 mem   0.05 GB tm     15.09us/   204.28ms (      0 GFLOPS    1|1      GB/s) 
*** CPU      229 E_250_4n1                                      arg  3 mem   0.05 GB tm     13.87us/   204.29ms (      0 GFLOPS    1|1      GB/s) ['__add__']
output: (1, 1000) 7
```


Caveats:
*  `test_flatten_reshape_add` (`b += 1` when `b = a.flatten()`) was the hardest case . `_try_simple_reshape_view_write` addresses this case but I would love to not need it.
* Diagonal & circular padding are still custom impl due to differences between torch and tinygrad. 
* I could not get batchnorm not fuse nicely during mnist training without the current impl. `aten::native_batch_norm` does more than `Tensor.batchnorm`.

Happy to hear if anyone has ideas how to make those three cases simpler.


Also: `torchbackendmore`/`torch_tests.py` fails silently (`|| true`) on an import error, the missing import is `expecttest` which is expected by `torch.testing._internal.common_utils` but not a declared dependency of pytorch. Should be fixed as a follow-up with an expectation on the expected failures. No idea if this ever worked as intended.

**Comment by roelofvandijk** [comment]
It did fuse when you added it in https://github.com/tinygrad/tinygrad/pull/9200, don't know when it broke (yet), bisecting...


Broke in April with this PR https://github.com/tinygrad/tinygrad/pull/9302.

The added realizes in the conv broke the fusion.

**Comment by geohot** [comment]
If tests pass, will merge and bounty is yours!

george@tinygrad.org to claim, can pay via PayPal or USDC on eth.
